lxd/storage/backend_lxd.go

```diff
@@ -226,6 +214,87 @@ func (b *lxdBackend) createInstanceSnapshotSymlink(inst Instance, mountPath stri
 	return nil
 }
 
+// removeInstanceSnapshotSymlinkIfUnused removes the symlink in the snapshot directory to the
+// instance's snapshot path if the snapshot path is missing. It is expected that the driver will
+// remove the instance's snapshot path after the last snapshot is removed or the volume is deleted.
+func (b *lxdBackend) removeInstanceSnapshotSymlinkIfUnused(inst Instance) error {
+	// Check we can convert the instance to the volume types needed.
+	volType, err := InstanceTypeToVolumeType(inst.Type())
+	if err != nil {
+		return err
+	}
+
+	snapshotSymlink := shared.VarPath("snapshots", project.Prefix(inst.Project(), inst.Name()))
+	volStorageName := project.Prefix(inst.Project(), inst.Name())
+
+	snapshotTargetPath, err := drivers.GetVolumeSnapshotDir(b.name, volType, volStorageName)
+	if err != nil {
+		return err
+	}
+
+	// If snapshot parent directory doesn't exist, remove symlink.
+	if !shared.PathExists(snapshotTargetPath) {
+		if shared.PathExists(snapshotSymlink) {
+			err := os.Remove(snapshotSymlink)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// CreateInstance creates an empty instance.
+func (b *lxdBackend) CreateInstance(inst Instance, op *operations.Operation) error {
+	logger := logging.AddContext(b.logger, log.Ctx{"project": inst.Project(), "instance": inst.Name()})
+	logger.Debug("CreateInstance started")
+	defer logger.Debug("CreateInstance finished")
+
+	volType, err := InstanceTypeToVolumeType(inst.Type())
+	if err != nil {
+		return err
+	}
+
+	revert := true
+	defer func() {
+		if !revert {
+			return
+		}
+		b.DeleteInstance(inst, op)
+	}()
+
+	vol := b.newVolume(volType, drivers.ContentTypeFS, project.Prefix(inst.Project(), inst.Name()), nil)
+	err = b.driver.CreateVolume(vol, nil, op)
+	if err != nil {
+		return err
+	}
+
+	err = b.createInstanceSymlink(inst, vol.MountPath())
+	if err != nil {
+		return err
+	}
+
+	// tomp TODO as this modifies the rootfs after being unpacked, it probably needs to run
```

Comment:
@stgraber I have a question about this, the `TemplateApply` function appears to modify the rootfs directory after it has been unpacked. This works fine for directory backends, but I'm concerned this won't work as it is for other backend drivers that need to mount. At first I thought I could move this into the filler func, but in the case of empty instances or instances created from optimised images (see below), then there is no filler func.

So should we pass a "templateFunc" into the driver's CreateVolume function that can be run after any filler funcs but before the volume is unmounted?

Reply:
@stgraber or perhaps instead we could modify the fillerFunc argument to be a slice that we can pass multiple functions to?

Reply:
done